In [4]:
pip install openpyxl

  Using cached openpyxl-3.1.5-py2.py3-none-any.whl.metadata (2.5 kB)
  Using cached et_xmlfile-2.0.0-py3-none-any.whl.metadata (2.7 kB)
Using cached openpyxl-3.1.5-py2.py3-none-any.whl (250 kB)
Using cached et_xmlfile-2.0.0-py3-none-any.whl (18 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [openpyxl]

[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [5]:
import pandas as pd

df = pd.read_excel("Analytics Case Study Dataset.xlsx")
df.head()

,Agmt Id,Cust Age,Cust Gender,Cust Cibil Score,Cust Employment Type,Cust Net Salary,Coborrower Flag,App Score Risk,Agmt Date,Seizure Date,...,OS Balance At Liquidation,Asset Age Months At Seizure,Months Spent In Yard,Asset Bodycondition,Asset Tyrecondition,Asset Generalcondition,Asset Enginecondition,Asset Accident Flag,Traiffic Challan Amount,Target Sold Amount At Liquidation
0,ASSET_1,29,F,-1,NREGI,30000,N,LOW RISK,2023-12-27,2025-05-23,...,46934.0,17.100000,1.033333,G,G,G,G,No,0,37000.0
1,ASSET_2,24,M,-1,NREGI,60000,N,LOW RISK,2025-11-20,2026-05-31,...,85177.0,6.400000,1.966667,G,G,G,G,No,0,56000.0
2,ASSET_3,32,F,763,AGR,29500,N,LOW RISK,2024-05-16,2025-06-24,...,51678.1,13.466667,2.066667,A,A,A,A,No,0,40500.0
3,ASSET_4,25,F,-1,NREGI,40000,N,MEDIUM RISK,2025-10-24,2026-04-30,...,65874.0,6.266667,1.900000,A,A,A,A,No,0,49000.0
4,ASSET_5,59,F,653,SAL,45000,N,MEDIUM RISK,2024-09-20,2025-05-30,...,66571.0,8.400000,1.833333,G,G,G,G,No,0,46000.0


In [6]:
print(df.columns.tolist())

['Agmt Id', 'Cust Age', 'Cust Gender', 'Cust Cibil Score', 'Cust Employment Type', 'Cust Net Salary', 'Coborrower Flag', 'App Score Risk', 'Agmt Date', 'Seizure Date', 'Sold Date', 'Tenure', 'Cust Net IRR', 'Sold Date Month', 'Sold Date Year', 'Seizure Date Month', 'Seizure Date Year', 'Cust Branch', 'Cust Region', 'Cust State', 'Pincode Tier', 'RC Availability', 'Registration Flag', 'Asset Disc Flag', 'Asset Alloy Flag', 'Asset Variant', 'Asset Model', 'Asset Fuel Type', 'Asset Cost At Disbursal', 'Loan Amount', 'LTV', 'OS Balance At Liquidation', 'Asset Age Months At Seizure', 'Months Spent In Yard', 'Asset Bodycondition', 'Asset Tyrecondition', 'Asset Generalcondition', 'Asset Enginecondition', 'Asset Accident Flag', 'Traiffic Challan Amount', 'Target Sold Amount At Liquidation']


In [9]:
print(df.shape)

(15000, 41)


In [10]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 41 columns):
 #   Column                             Non-Null Count  Dtype         
---  ------                             --------------  -----         
 0   Agmt Id                            15000 non-null  str           
 1   Cust Age                           15000 non-null  int64         
 2   Cust Gender                        15000 non-null  str           
 3   Cust Cibil Score                   15000 non-null  int64         
 4   Cust Employment Type               15000 non-null  str           
 5   Cust Net Salary                    15000 non-null  int64         
 6   Coborrower Flag                    15000 non-null  str           
 7   App Score Risk                     15000 non-null  str           
 8   Agmt Date                          15000 non-null  datetime64[us]
 9   Seizure Date                       15000 non-null  datetime64[us]
 10  Sold Date                          15000 non-

In [11]:
df.isnull()

,Agmt Id,Cust Age,Cust Gender,Cust Cibil Score,Cust Employment Type,Cust Net Salary,Coborrower Flag,App Score Risk,Agmt Date,Seizure Date,...,OS Balance At Liquidation,Asset Age Months At Seizure,Months Spent In Yard,Asset Bodycondition,Asset Tyrecondition,Asset Generalcondition,Asset Enginecondition,Asset Accident Flag,Traiffic Challan Amount,Target Sold Amount At Liquidation
0,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14995,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
14996,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
14997,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
14998,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False


In [12]:
df.isnull().sum().sort_values(ascending=False)

Agmt Id                              0
RC Availability                      0
Asset Disc Flag                      0
Asset Alloy Flag                     0
Asset Variant                        0
Asset Model                          0
Asset Fuel Type                      0
Asset Cost At Disbursal              0
Loan Amount                          0
LTV                                  0
OS Balance At Liquidation            0
Asset Age Months At Seizure          0
Months Spent In Yard                 0
Asset Bodycondition                  0
Asset Tyrecondition                  0
Asset Generalcondition               0
Asset Enginecondition                0
Asset Accident Flag                  0
Traiffic Challan Amount              0
Registration Flag                    0
Pincode Tier                         0
Cust Age                             0
Cust State                           0
Cust Gender                          0
Cust Cibil Score                     0
Cust Employment Type     

In [13]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 41 columns):
 #   Column                             Non-Null Count  Dtype         
---  ------                             --------------  -----         
 0   Agmt Id                            15000 non-null  str           
 1   Cust Age                           15000 non-null  int64         
 2   Cust Gender                        15000 non-null  str           
 3   Cust Cibil Score                   15000 non-null  int64         
 4   Cust Employment Type               15000 non-null  str           
 5   Cust Net Salary                    15000 non-null  int64         
 6   Coborrower Flag                    15000 non-null  str           
 7   App Score Risk                     15000 non-null  str           
 8   Agmt Date                          15000 non-null  datetime64[us]
 9   Seizure Date                       15000 non-null  datetime64[us]
 10  Sold Date                          15000 non-

In [14]:
target = "Target Sold Amount At Liquidation"

df[target].describe()

count     15000.000000
mean      44689.349355
std       16602.006132
min       10000.000000
25%       33000.000000
50%       43000.000000
75%       55000.000000
max      170000.000000
Name: Target Sold Amount At Liquidation, dtype: float64

First Observation
Mean   = 44,689

Median = 43,000
Mean and median are very close

Target is not heavily skewed

✅ CatBoost can directly predict target

Second Observation

Range

10,000
to
170,000

That's a huge spread.

But:

75% = 55,000

Meaning most assets are below:

₹55,000

while a small number go much higher.

This suggests a few premium vehicles exist.

Third Observation

Standard deviation

16,602

Relative to mean:

16602 / 44689

≈

37%

Meaning:

Residual value varies significantly

Good news.

There's enough variation for ML to learn patterns.

Business Interpretation

Typical repossessed vehicle:

Expected Recovery

≈ ₹45,000

This Gives Us a Baseline

If your model predicts:

₹45,000

for every vehicle,

RMSE will already be somewhat decent.

Your model must significantly beat that.

In [15]:
numeric_cols = df.select_dtypes(include=["int64","float64"]).columns

corr = df[numeric_cols].corr()["Target Sold Amount At Liquidation"]

print(corr.sort_values(ascending=False))

Target Sold Amount At Liquidation    1.000000
Asset Cost At Disbursal              0.660987
Loan Amount                          0.660977
OS Balance At Liquidation            0.629471
Asset Disc Flag                      0.291538
Tenure                               0.267396
Months Spent In Yard                 0.077037
Asset Alloy Flag                     0.072366
Cust Net Salary                      0.037613
Cust Cibil Score                     0.026671
LTV                                 -0.006644
Seizure Date Year                   -0.011794
Sold Date Month                     -0.012888
Sold Date Year                      -0.014860
Seizure Date Month                  -0.049890
Cust Age                            -0.085409
Traiffic Challan Amount             -0.096437
Cust Net IRR                        -0.150355
Asset Age Months At Seizure         -0.395156
Name: Target Sold Amount At Liquidation, dtype: float64


Biggest Drivers of Residual Value
Feature	                     Correlation
Asset Cost At Disbursal	      0.661
Loan Amount	                  0.661
OS Balance At Liquidation	  0.629
Asset Disc Flag	              0.292
Tenure	                      0.267
Asset Age Months At Seizure	  -0.395

Business Insight 1
Asset Cost = King
Asset Cost At Disbursal = 0.661

Makes perfect sense.

Expensive vehicles:

Higher resale value
Higher recovery

This will likely become the #1 SHAP feature.

Business Insight 2
Asset Age Is Your Biggest Risk Driver
Asset Age Months At Seizure = -0.395

Translation:

Older vehicle
↓
Lower resale value
↓
Higher residual risk

This is exactly what TVS wants to understand.

Business Insight 3
Outstanding Balance Matters
OS Balance = 0.629

Interesting.

This means assets with larger outstanding balances tend to have higher liquidation values.

Probably because:

Higher value vehicle
↓
Higher financed amount
↓
Higher outstanding balance

Business Insight 4
Months In Yard Is Surprisingly Weak
0.077

I expected stronger.

Possible reasons:

Most assets spend similar time in yard
Dataset is synthetic
Effect already captured by age/condition

This is why EDA matters.

Business Insight 5
LTV Almost Doesn't Matter
-0.006

This shocked me.

It means:

LTV doesn't directly affect sale value

Which is actually logical.

LTV affects:

Risk
Profitability
Loss

Not the market value of the vehicle.

In [16]:
condition_map = {
    "G":3,
    "A":2,
    "P":1
}

In [18]:
df[["Asset_Health_Index",
    "Target Sold Amount At Liquidation"]].corr()

,Asset_Health_Index,Target Sold Amount At Liquidation
Asset_Health_Index,1.000000,0.211404
Target Sold Amount At Liquidation,0.211404,1.000000


Huge Hackathon Insight

Notice something:

The problem statement asks for:

Residual Risk Score
Profitability Score
Recovery Efficiency Index
Asset Health Index

But the dataset does NOT provide them.

That means:

TVS expects participants to engineer these features.

This is where many teams lose points.

In [19]:
for col in [
    "Asset Bodycondition",
    "Asset Tyrecondition",
    "Asset Generalcondition",
    "Asset Enginecondition"
]:
    print("\n", col)
    print(df[col].value_counts())


 Asset Bodycondition
Asset Bodycondition
G    7488
A    6600
P     912
Name: count, dtype: int64

 Asset Tyrecondition
Asset Tyrecondition
G    7498
A    6589
P     913
Name: count, dtype: int64

 Asset Generalcondition
Asset Generalcondition
G    7749
A    6511
P     740
Name: count, dtype: int64

 Asset Enginecondition
Asset Enginecondition
G    8129
A    6179
P     692
Name: count, dtype: int64


First Business Insight

Engine condition has the highest number of Good assets.

Engine Good = 8129

Poor engine condition is relatively rare.

Engine Poor = 692

This usually means:

Poor Engine
↓
Huge recovery impact

because it's uncommon and likely expensive to repair.

##Build Asset Health Index

Why?

Engine > General > Body > Tyre

Engine condition affects resale value much more than tyres.

This sounds smarter in a presentation too.

In [20]:
condition_map = {
    "G": 100,
    "A": 70,
    "P": 30
}

df["Asset_Health_Index"] = (
      0.35 * df["Asset Enginecondition"].map(condition_map)
    + 0.25 * df["Asset Generalcondition"].map(condition_map)
    + 0.20 * df["Asset Bodycondition"].map(condition_map)
    + 0.20 * df["Asset Tyrecondition"].map(condition_map)
)

In [21]:
df[["Asset_Health_Index",
    "Target Sold Amount At Liquidation"]].corr()

,Asset_Health_Index,Target Sold Amount At Liquidation
Asset_Health_Index,1.000000,0.217872
Target Sold Amount At Liquidation,0.217872,1.000000


Asset Cost and Asset Age are primary drivers of residual value, while Asset Health contributes additional explanatory power and improves interpretability of lending decisions.

##Build Recovery Efficiency Index

In [23]:
df["Recovery_Efficiency_Index"] = (
    df["Target Sold Amount At Liquidation"]
    /
    df["OS Balance At Liquidation"]
) * 100

##Build LGD

Loss Given Default.

In [24]:
df["LGD"] = (
    df["OS Balance At Liquidation"]
    -
    df["Target Sold Amount At Liquidation"]
)

In [25]:
df["LGD_pct"] = (
    (
        df["OS Balance At Liquidation"]
        -
        df["Target Sold Amount At Liquidation"]
    )
    /
    df["OS Balance At Liquidation"]
) * 100

In [26]:
from sklearn.preprocessing import MinMaxScaler

risk_features = pd.DataFrame({
    "age": df["Asset Age Months At Seizure"],
    "lgd": df["LGD_pct"],
    "challan": df["Traiffic Challan Amount"],
    "health": 100 - df["Asset_Health_Index"]
})

scaler = MinMaxScaler()

scaled = scaler.fit_transform(risk_features)

df["Residual_Risk_Score"] = (
    scaled[:,0]*0.35 +
    scaled[:,1]*0.35 +
    scaled[:,2]*0.10 +
    scaled[:,3]*0.20
) * 100

In [27]:
def risk_band(x):
    if x <= 25:
        return "Low"
    elif x <= 50:
        return "Medium"
    elif x <= 75:
        return "High"
    else:
        return "Critical"

df["Risk_Band"] = df["Residual_Risk_Score"].apply(risk_band)

What We Have After This

Your Digital Twin now contains:

Agreement

Asset Health Index

Recovery Efficiency Index

LGD

Residual Risk Score

Risk Band

In [28]:
df[["Asset_Health_Index",
    "Target Sold Amount At Liquidation"]].corr()

,Asset_Health_Index,Target Sold Amount At Liquidation
Asset_Health_Index,1.000000,0.217872
Target Sold Amount At Liquidation,0.217872,1.000000


In [29]:
df["Risk_Band"].value_counts()

Risk_Band
Medium      11529
High         3459
Critical        9
Low             3
Name: count, dtype: int64